packages/template-retail-react-app/app/utils/url.js

```diff
@@ -150,22 +151,77 @@ export const getPathWithLocale = (shortCode, opts = {}) => {
+        isDefaultLocaleOfDefaultSite && isDefaultSite && isHomeRef
+            ? ''
+            : siteRef || defaultSite.alias || defaultSite.id,
+        isDefaultLocaleOfDefaultSite && isDefaultSite && isHomeRef ? '' : shortCode
     )
     return newUrl
 }
 
+/**
+ * Generates the URL Template literal (Template string) used to build URLs in the App according
+ * the current selected site/locale and the default App URL configuration.
+ *
+ * @param appConfig Application default configuration.
+ * @param siteUrl Current selected Site object.
+ * @param localeUrl Current selected Locale object.
+ * @returns {function(*, *, *, *=): string} function providing href, site and locale generates a site URL.
+ */
+export const getUrlTemplateLiteral = (appConfig, siteUrl, localeUrl) => {
+    const {site: siteConfig, locale: localeConfig, showDefaults: showDefaultsConfig} = appConfig.url
+    const defaultSite = getDefaultSite()
+    const sites = getSites()
+    const site =
+        sites.find((site) => {
+            return site.alias === appConfig['site'] || site.id === appConfig['site']
+        }) || defaultSite
+    const defaultLocale = getLocaleByReference(site, site.l10n.defaultLocale)
+
+    const isDefaultSite = defaultSite === siteUrl
+    const isDefaultLocale = defaultLocale === localeUrl
+
+    const querySite =
+        (siteConfig === urlPartPositions.QUERY_PARAM && showDefaultsConfig) ||
+        (siteConfig === urlPartPositions.QUERY_PARAM && !showDefaultsConfig && !isDefaultSite)
+    const queryLocale =
+        (localeConfig === urlPartPositions.QUERY_PARAM && showDefaultsConfig) ||
+        (localeConfig === urlPartPositions.QUERY_PARAM && !showDefaultsConfig && !isDefaultLocale)
+
+    const isQuery = querySite || queryLocale
+    const isQuerySiteAndLocale = querySite && queryLocale
+
+    const pathSite =
+        (siteConfig === urlPartPositions.PATH && showDefaultsConfig) ||
+        (siteConfig === urlPartPositions.PATH && !showDefaultsConfig && !isDefaultSite)
+    const pathLocale =
+        (localeConfig === urlPartPositions.PATH && showDefaultsConfig) ||
+        (localeConfig === urlPartPositions.PATH && !showDefaultsConfig && !isDefaultLocale)
+
+    const isPathSiteAndLocale = pathSite && pathLocale
+
+    const templateConfig = {
+        pathSite: pathSite,
+        pathLocale: pathLocale,
+        isPathSiteAndLocale: isPathSiteAndLocale,
+        isQuery: isQuery,
+        isQuerySiteAndLocale: isQuerySiteAndLocale,
+        querySite: querySite,
+        queryLocale: queryLocale
+    }
+
+    return (href, site, locale, c = templateConfig) =>
+        `${c.pathSite && site ? `/${site}` : ''}${c.pathLocale && locale ? `/${locale}` : ''}` +
+        `${href}` +
+        `${c.isQuery && (site || locale) ? '?' : ''}${c.querySite && site ? `site=${site}` : ''}${
+            c.isQuerySiteAndLocale && locale ? '&' : ''
+        }${c.queryLocale && locale ? `locale=${locale}` : ''}`
```

Comment:
This is a pretty bulky template string. It might improve readability to split it up into a couple of `if` blocks.

Reply:
Agreed with Will on that. ☝️